Sources/Datadog/Core/Upload/RequestBuilder.swift

```diff
@@ -126,9 +127,25 @@ internal struct RequestBuilder {
         computedHeaders.forEach { field, value in headers[field] = value() }
 
         request.httpMethod = "POST"
-        request.allHTTPHeaderFields = headers
-        request.httpBody = data
 
+        if let body = Deflate.encode(data) {
+            headers[HTTPHeader.contentEncodingHeaderField] = "deflate"
+            request.httpBody = body
+        } else {
+            request.httpBody = data
+#if DD_SDK_ENABLE_INTERNAL_MONITORING
+            InternalMonitoringFeature.instance?.monitor.sdkLogger.warn(
```

Comment:
`RequestBuilder` is [also used by the `InternalMonitoringFeature` itself](https://github.com/DataDog/dd-sdk-ios/blob/fcbacd6f7a2cfaaa9159cb6140a5ca19453f492f/Sources/Datadog/InternalMonitoring/InternalMonitoringFeature.swift#L69). To avoid infinite loop in IM (_"send internal log → report compression warning → send internal log → ..."_ we should inject the `monitor: InternalMonitor?` as a dependency to `RequestBuilder` and `nil-ify` it for the `InternalMonitoringFeature`. Similar to [what we do in `DataUploadWorker`](https://github.com/DataDog/dd-sdk-ios/blob/fcbacd6f7a2cfaaa9159cb6140a5ca19453f492f/Sources/Datadog/Core/Upload/DataUploadWorker.swift#L28-L29).